### PR TITLE
Fix #25398: Template Builder Disable Row When a Box Cannot Fit

### LINKED ISSUE #25398: Template Builder: End user should be notified in case you are unable to drag and drop boxes into the row
### Problem Statement

As an end user of dotCMS, I want to be notified if I am unable to drag and drop a box into the template designer so that I can understand why it is not possible and find an alternative solution.

https://github.com/dotCMS/core/assets/2641437/c0c3f2b9-8c1c-487d-8dbb-91d8c78650dc

### Steps to Reproduce

- Log in to dotCMS.
- Go to the template designer.
- Attempt to drag and drop more than 4 boxes into the template designer.

### Acceptance Criteria

- When attempting to drag and drop a box into the template designer, the system should provide a clear error message indicating the reason why it is not possible.
- The error message should provide guidance on alternative solutions or actions to be taken.
- The error message should be displayed prominently and clearly visible to the end user.

### dotCMS Version

Tested on master // Docker // FF 

### Proposed Objective

Quality Assurance

### Proposed Priority

Priority 2 - Important

### PR DESCRIPTION
**Pull request by zJaaal**
### Summary
🚀🧮🎨

Improved the layout logic and performance of the template builder components. Added a `willBoxFit` property to the gridstack items and a utility function to check the available space in a row. Simplified the imports and change detection of some components. Added or updated tests and mocks for the new functionality.

> _Sing, O Muse, of the skillful code review_
> _That improved the template builder's design_
> _And made the gridstack boxes fit and shine_
> _Like stars in the clear sky of Zeus above._

### Walkthrough
*  Add `willBoxFit` property to `DotGridStackWidget` and `DotGridStackNode` interfaces to indicate if a box can fit in a row ([link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-b93c6205d24c66efa12727a86a6616d0bd0fdcd030871cd4ead5d72a69792235R41),[link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-b93c6205d24c66efa12727a86a6616d0bd0fdcd030871cd4ead5d72a69792235R56))
*  Implement `willBoxFitInRow` function to determine if a box can fit in a row based on gridstack coordinates and dimensions ([link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-f3792a46b295fb618b6d4a8d8f4ecef72ec5e3c41a470d6a44886881a6b93659R198-R219))
*  Use `willBoxFitInRow` function to compute `willBoxFit` property for each item in the template builder store ([link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-a32a0fcbcf1b026389eade1b0cd88e8838bbfacb92905fe9da2bea7f72b7c9abL22-R23),[link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-a32a0fcbcf1b026389eade1b0cd88e8838bbfacb92905fe9da2bea7f72b7c9abL37-R44))
*  Modify `dotcms-template-builder-row` component to use `willBoxFit` property to adjust the opacity of the background columns ([link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-06b8e2225be93160f82e1235377e9c34cdd7d99de753b73bdc3d0abf70b42192L28-R28),[link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-6cb544636b5cb7d58af6da435272da8e73232f0cb288ea1b4eb90edd81c14e99L4-R4),[link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-6cb544636b5cb7d58af6da435272da8e73232f0cb288ea1b4eb90edd81c14e99L24))
*  Add test cases for `willBoxFitInRow` function using mock data ([link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-3647ec392c96563c5712834ab3a47c47021cc2e0b7aa88a8add5bd49849a6a42L3-R8),[link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-3647ec392c96563c5712834ab3a47c47021cc2e0b7aa88a8add5bd49849a6a42R76-R101),[link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-98de89ac988dbfd17c912bb493096714d7c139835179a99162b4df2eb3c309e5R423-R617))
*  Add test case for `dotcms-template-builder-background-columns` component to check the background when a box can fit in a row ([link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-0d94175e271fba9576c6032c57ca4c5641c0a80cb6bc9ebd8dfcc3a7f607e90aL33-R34),[link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-0d94175e271fba9576c6032c57ca4c5641c0a80cb6bc9ebd8dfcc3a7f607e90aR104-R115))
*  Update `template-builder.store.spec.ts` to use direct selectors instead of `pluck` operator ([link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-31b7a396b598816d0a94599edd0c1338d80f276fed9775642769710040776e24L54-R55))
*  Replace `DotMessagePipeModule` import by `DotMessagePipe` in `dot-device-selector-seo.component.stories.ts` to avoid importing the whole module ([link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-9d9dcbd8c15376938e7e220789fec014b153dad1c1a2e2e5703ac063c7f213c2L18-R18),[link](https://github.com/dotCMS/core/pull/25513/files?diff=unified&w=0#diff-9d9dcbd8c15376938e7e220789fec014b153dad1c1a2e2e5703ac063c7f213c2L52-R52))

### Screenshots
https://github.com/dotCMS/core/assets/63567962/1f847f49-65ab-4a72-b7fc-0205881533fc

### CODE REVIEW

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts [hunk 1]
```diff
@@ -195,3 +195,25 @@ export function getBoxVariantByWidth(width: number): TemplateBuilderBoxSize {
...
+ *
+ * @export
+ * @param {DotGridStackWidget} row
+ * @return {*} boolean
+ */
+export function willBoxFitInRow(row: DotGridStackWidget): boolean {
+    const emptyChar = '_';
+    const boxChar = '#';
+
+    if (!row.subGridOpts) return false;
```
Comment (jdotcms) [inline core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts:210]: always always use curly brackets on a conditions

**Review by jdotcms — APPROVED** [review]

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts [hunk 2]
```diff
@@ -189,9 +191,37 @@ export const parseFromGridStackToDotObject = (gridData: DotGridStackWidget[]): D
...
+        return false;
+    }
+
+    const rowSpace = Array.from({ length: 12 }, () => emptyChar); // Array with 12 empty spaces
+
+    row.subGridOpts.children.forEach(({ x, w }) => {
+        rowSpace.splice(x, w, ...boxChar.repeat(w).split('')); // Fill needed empty spaces with #
+    });
+
+    return /___/.test(rowSpace.join('')); // If there are 3 consecutive empty spaces then we can drop one more box
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts:226]: If we change the width of the new box we will have to remember that we need to change it here... can we create somehow a constant that we can just change in one place?

Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts:226]: If we change the width of the new box we will have to remember that we need to change it here... can we create somehow a constant that we can just change in one place?

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.html [hunk 3]
```diff
@@ -50,7 +50,7 @@
             <dotcms-template-builder-row
                 class="grid-stack-item"
                 #rowElement
-                *ngFor="let row of vm.items; trackBy: identify"
+                *ngFor="let row of vm.rows; trackBy: identify"
```
Comment (fmontes) [inline core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.html:53]: Guessing will need the class here?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts (hunk 2 above)
Comment (zJaaal) [reply core-web/libs/template-builder/src/lib/components/template-builder/utils/gridstack-utils.ts:226]: Of course, working on it!

File: core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.html (hunk 3 above)
Comment (zJaaal) [reply core-web/libs/template-builder/src/lib/components/template-builder/template-builder.component.html:54]: Yes, working on it!

**Review by rjvelazco — APPROVED** [review]

**Comment by zJaaal** [comment]
## Update
We changed the approach to the one in the video.

https://github.com/dotCMS/core/assets/63567962/4e0ec445-ba7f-486d-a60a-670c5f4a512d

**Review by danielmdob — APPROVED** [review]
LGTM!

**Comment by zJaaal** [comment]
Leaving in draft until we find out how to work around the issue explained in: `core-web/libs/template-builder/src/lib/components/template-builder/components/add-widget/add-widget.component.ts` at Line 40.

**Review by fmontes — APPROVED** [review]

**Pull request merged**